In [46]:
import random
import numpy as np

In [85]:
from bokeh.io import show
from bokeh.plotting import figure
from bokeh.io import output_notebook
from bokeh.layouts import gridplot

Often, in big data, we are asked to tell the difference between two samples. Was there really more rain in 2018 than in 2019? Was the temperature hotter in these two years? Did more people ride bikes in Boston than in Seattle? 

There is a mathmatical way in big data where we can answer these questions. 

Imagine a small town.There are two schools. Each school has two classrooms. In the north school, the two classrooms have students who are 12 years old. In the south class room, classroom a has students who are 12 years old, but classroom b has students who are 17 years old. 

Answer the following:
Do you expect a difference in height between the students in the north school (both classrooms has students who are 12 years old)? Why? 

Do you expect a difference in height between the students in the two classrooms in the south school (one class has students who are 12 years old, and one has students who are 17 years old)? Why?

In [35]:
def get_classroom_north_a():
    return [146.1,152.7, 146.3, 142.0, 151.8,
     151.4, 145.1, 153.5, 151.2, 143.5, 158.2,
     150.6, 143.5, 151.6, 149.9, 154.2, 142.4,
     154.6, 154.1, 152.8, 152.4, 155.9, 152.9,
     149.9, 145.0]


In [75]:
def get_classroom_north_b():
    return [142.9, 150.9, 154.0, 146.4, 142.4,
     148.7, 151.4, 154.5, 142.9, 142.5, 152.5,
     151.4, 156.7, 153.9, 148.5, 147.6, 161.9,
     147.6, 145.1, 143.3, 149.5, 147.3,
     148.7, 150.4]

In [81]:
def get_classroom_south_a():
    return [141.8, 150.2, 147.6, 146.6, 153.8,
     149.3, 147.6, 158.0, 146.5, 142.7, 142.1,
     146.4, 152.3, 153.3, 154.7, 158.3, 157.6,
     152.3, 155.8, 152.4, 146.4, 153.3, 149.5,
     148.2, 159.3]

In [66]:
def get_classroom_south_b():
    return [169.6, 163.3,  177.6, 164.5,
     169.5, 168.9, 168.4, 168.7, 163.3, 163.4,
     165.0, 164.0, 169.9, 173.6, 161.3, 168.5,
     160.8, 162.3, 164.6, 166.3, 163.6, 152.6,
     172.8, 164.4]

The above functions has the heights for the students in each classroom. Lets graph the 4 classrooms.

In [126]:
NORTH_A = get_classroom_north_a()
NORTH_B = get_classroom_north_b()
SOUTH_A = get_classroom_south_a()
SOUTH_B = get_classroom_south_b()

In [140]:
def make_bar(labels, nums, title = None, y_range = None, plot_width = 350, plot_height = 350):
    p = figure(title = title, plot_width = plot_width, plot_height = plot_height,
              y_range = y_range)
    p.vbar(x=labels, top=nums, width=0.9)
    p.xgrid.grid_line_color = None
    return p

P_NORTH_A = make_bar(labels = [x for x in range(len(NORTH_A))], nums = sorted(NORTH_A), 
               title = 'North A', y_range = (120, 160))
P_NORTH_B = make_bar(labels = [x for x in range(len(NORTH_B))], nums = sorted(NORTH_B), 
               title = "North B", y_range = (120, 160))
P_SOUTH_A = make_bar(labels = [x for x in range(len(SOUTH_A))], nums = sorted(SOUTH_A), 
                title = 'South A', y_range = (120, 180))
P_SOUTH_B = make_bar(labels = [x for x in range(len(SOUTH_B))], nums = sorted(SOUTH_B), 
                title = "South B", y_range = (120, 180))


grid = gridplot([ P_NORTH_A, P_NORTH_B, P_SOUTH_A,P_SOUTH_B,], ncols = 2)
show(grid)

Let's get the means

In [128]:
print(np.mean(SOUTH_A))
print(np.mean(SOUTH_B))
print(np.mean(NORTH_A))
print(np.mean(NORTH_B))


150.64
166.12083333333337
150.064
149.20833333333334


In [111]:
output_notebook()

Loading BokehJS ...

For the south school, classroom a has a mea of 150.064, and classroom b has a mean of 149.208. That is a difference of .86 cm. 

Answer the following:

Do you still think there is no difference between the two classrooms in the north school? Why? 

Mabye the difference in heigh between the two classrooms in the north school is just due to randomness. How can we know? 

Imagine an experiment. We got to a big city and find 1,000 students who are 12 years old. We ramdonly choose 25 for the first class, and 24 for the second (the original number of students for each). We calcuate the mean for each classroom. 

Answer the following: Do you expect the same result as the mean above (150.064 and 149.228)? Why? Do you expect the mean from either group to be much different from the original (around 150 cm)? Why? 

Now, imagine that we conduct the same experiment in the south classroom. We go into a big city and find 1,000 students who are 12 year olds. We choose 25 for class a. We find 1,000 students who are 17 years old. We choose 24 for classroom b. We calculate the mean for each classroom

Answer the following:

Do you expect the means for the two classrooms to be different? (Remember, on classroom has old students, and one has young students.) Why? 

Let's conduct the experiment in Python

In [400]:
def resample(l):
    """There is a better way to do this; but this helps in explaining"""
    population = l * 40 # we are making around 1,000 studens
    random.shuffle(population) # mix them up
    return population[:len(l)] # return number of sample size

def make_grid():
    resample_a = resample(NORTH_A)
    resample_b = resample(NORTH_B)
    p1= make_bar(labels = [x for x in range(len(resample_a))], nums = sorted(resample_a), 
               title = "Resample A", y_range = (120, 160))
    p2 = make_bar(labels = [x for x in range(len(resample_b))], nums = sorted(resample_b), 
               title = "Resample B", y_range = (120, 160))
    grid = gridplot([p1, p2], ncols = 2)
    print('the mean of resample 1 is {r1} and the mean of resample2 is {r2}'.format(
        r1 = np.mean(resample_a), r2 = np.mean(resample_b)))
    return grid
show(make_grid())


the mean of resample 1 is 149.92 and the mean of resample2 is 149.12083333333334


The result is random, but you will most likely see that the graphs don't look different. Now let's try the same experiment for the south school


In [396]:
import math
def resample(l):
    final = []
    for i in range(len(l)):
        final.append(random.choice(l))
    return final

means = []
n = 10000
o = [random.gauss(0, 1) for x in range(1000)]
t =np.mean(o) +  (np.std(o)/math.sqrt(1000)) * 2
print(t)

for i in range(n):
    means.append(np.mean(resample_(o)))

print(np.quantile(means, .95))



0.10977758524985742
0.09885611109761742


In [157]:
BOTH_A_B = SOUTH_A + SOUTH_B
random.shuffle(BOTH_A_B)
NEW_SAMPLE = BOTH_A_B[0:25] 

P_NEW = make_bar(labels = [x for x in range(len(NEW_SAMPLE))], nums = sorted(NEW_SAMPLE), 
               title = "New Sample", y_range = (120, 180))
GRID = gridplot([P_SOUTH_A, P_NEW], ncols = 2)
print('The mean of the origin sample is {o} and the mean of the resampled is {r}'.format(
    o = np.mean(SOUTH_A), r = np.mean(NEW_SAMPLE))
     )
print(np.mean(SOUTH_A) - np.mean(SOUTH_B))
show(GRID)

The mean of the origin sample is 150.64 and the mean of the resampled is 159.424
-15.48083333333338


In [154]:
def resample_two_samples(sample1, sample2, num_iterations = 100):
    both = sample1 + sample2
    mean1 = []
    mean2 = []
    for i in range(num_iterations):
        random.shuffle(both) 
        new_1 = both[0:len(sample1)] 
        new_2 = both[len(sample1):] 
        mean1.append(np.mean(new_1)) 
        mean2.append(np.mean(new_2)) 
    return mean1, mean2

In [161]:
resampled_means1, resampled_means2 = resample_two_samples(get_classroom_north_a(), get_classroom_north_b())
resampled_means1 = [round(x,1) for x in resampled_means1]
resampled_means2 = [round(x,1) for x in resampled_means2]
print(resampled_means1)
print()
print(resampled_means2)

[149.2, 150.2, 149.4, 149.2, 150.5, 150.4, 149.3, 148.6, 150.8, 149.7, 149.7, 149.7, 150.9, 149.5, 149.8, 149.7, 149.7, 150.2, 149.4, 149.9, 149.6, 148.3, 149.8, 149.8, 149.4, 149.8, 150.0, 149.1, 149.3, 148.8, 150.1, 149.2, 149.9, 149.4, 151.4, 150.1, 149.5, 149.9, 149.1, 150.3, 150.4, 150.2, 149.3, 148.8, 149.1, 149.4, 149.1, 150.5, 149.6, 149.2, 150.1, 149.3, 149.8, 150.0, 150.3, 149.6, 150.1, 151.3, 149.5, 150.0, 149.9, 148.9, 150.2, 148.6, 149.0, 151.0, 150.0, 149.6, 149.0, 150.7, 148.9, 149.5, 151.1, 149.6, 149.1, 149.4, 150.5, 148.4, 149.9, 149.6, 149.5, 149.6, 148.8, 150.4, 150.1, 150.7, 149.0, 149.9, 150.2, 149.6, 150.3, 150.0, 150.3, 150.0, 150.2, 148.8, 149.2, 149.0, 150.3, 149.5]

[150.1, 149.1, 149.9, 150.1, 148.8, 148.8, 150.0, 150.7, 148.4, 149.6, 149.6, 149.6, 148.4, 149.8, 149.5, 149.6, 149.6, 149.1, 149.9, 149.4, 149.7, 151.1, 149.4, 149.4, 149.9, 149.5, 149.3, 150.3, 150.0, 150.6, 149.2, 150.1, 149.4, 149.9, 147.8, 149.2, 149.8, 149.4, 150.2, 149.0, 148.9, 149.0, 150

Have a look at the code above. We used Python to conduct the experiment with bringing the students into the cafeteria, and randomly choosing 25. The random.shuffle code mixed the students. We did this experiment 100 times. Each time we did the experiment, we calculated the mean. 

Each of these samples is called a *resampled* mean because we re sampled (in this case, 100 times).

Answer the following. Do not use a calculator! The mean of classroom a was 150. Does the first resampeld mean look any different from that mean? Why?

The mean for classroom b was 149. Does the mean of the second sample look any different from the the resampled mean? Why? 

In [56]:
resampled_means1, resampled_means2 = resample_two_samples(get_classroom_south_a(), get_classroom_south_b())
resampled_means1 = [round(x,1) for x in resampled_means1]
resampled_means2 = [round(x,1) for x in resampled_means2]
print(resampled_means1)
print()
print(resampled_means2)

[157.5, 157.8, 157.9, 158.7, 159.3, 157.0, 155.7, 158.3, 156.8, 158.3, 156.7, 158.5, 159.9, 161.4, 158.8, 159.7, 156.7, 157.5, 157.0, 160.8, 157.6, 158.5, 158.3, 158.8, 158.2, 156.5, 158.3, 159.3, 156.9, 158.7, 158.8, 158.5, 156.5, 158.1, 159.4, 160.1, 158.9, 160.1, 157.2, 157.7, 159.1, 159.0, 158.5, 159.3, 157.0, 157.9, 157.4, 158.4, 159.8, 159.3, 160.6, 159.1, 157.7, 159.1, 156.0, 156.5, 159.1, 156.6, 159.1, 158.5, 157.7, 158.9, 159.6, 156.8, 155.6, 159.0, 156.2, 161.8, 160.7, 160.2, 160.8, 159.8, 159.1, 156.1, 159.8, 157.4, 159.1, 159.9, 158.6, 159.0, 158.4, 158.3, 159.7, 158.1, 157.5, 160.8, 156.4, 161.8, 157.7, 158.3, 156.9, 159.0, 159.3, 159.5, 159.0, 160.9, 158.3, 157.9, 160.0, 160.0]

[159.2, 159.0, 158.9, 158.1, 157.4, 159.8, 161.1, 158.4, 159.9, 158.4, 160.1, 158.2, 156.8, 155.3, 157.9, 157.0, 160.1, 159.3, 159.7, 155.9, 159.1, 158.3, 158.4, 157.9, 158.5, 160.2, 158.4, 157.5, 159.9, 158.1, 157.9, 158.2, 160.2, 158.6, 157.3, 156.7, 157.8, 156.6, 159.5, 159.0, 157.7, 157.7, 158

We resampled the two classrooms from the south school

Answer the following. Does the mean of classroom a (151) look different from the resample a?

Does the the mean of classroom b (166) look different from the resample b? 

Population vs Samples

In statistics, we talk about a sample and the population. The population is *everthing*. For example, the population for the average rainfall in the Amazon is the average for every single year, from the beginnig of time until the end of the world. A sample is the set of year 2000 to 2010. 

Generally we think of our data sets as samples. We can never have all the data. We only have little bits. We have to guess what these little bits mean.

In the above example with students, we can think of each classroom as a sample. We can think of mixing the students as the population. 

If the students come from the same population, then if we mix them up (resample) we would expect to get the same results. But if they came from different populations, we would expect to get different results. 

For the south school, we can think of the students coming from the same population. That is why the resampled means look the same.

But for the north school, the students came from different populations. 

In [62]:
# redo the experiment
resampled_means1, resampled_means2 = resample_two_samples(get_classroom_north_a(), get_classroom_north_b())
diff1 = []
#get the mean of classroom a (150.1)
mean1 = np.mean(get_classroom_north_a())
# find out how different each resampled mean is from the classroom
for i in resampled_means1:
    difference = mean1 - i
    difference = round(difference,1)
    diff1.append(difference)
print(diff1)


[0.5, -0.8, 0.4, -0.4, -0.3, -0.4, 0.5, 0.2, -0.3, 0.7, 0.8, 1.1, 0.8, 0.9, 0.5, 1.0, 0.1, 0.3, 0.3, 0.2, 0.4, 0.9, 0.3, 1.3, 0.7, 0.1, 0.2, 0.3, 0.3, -0.2, 0.7, -0.3, 0.9, 0.1, 0.4, 1.3, -0.2, -0.3, -0.2, 0.5, 0.8, 0.4, 0.7, 1.4, -0.3, -0.3, 0.8, -1.0, -0.5, 0.5, 0.9, -0.5, 0.2, -0.5, 0.3, 0.5, 1.4, 0.2, 1.8, 1.0, -0.4, 0.8, 0.6, 0.9, 1.5, 1.2, -0.4, 0.6, 1.5, 0.3, 1.2, 0.2, -1.2, 0.1, 0.3, 0.4, -0.4, -0.1, 0.4, 2.3, 0.3, -0.5, 0.3, 0.7, -0.5, 0.1, -0.9, 0.3, 0.6, 0.8, 0.3, 0.8, -0.0, 0.7, 0.7, 0.4, -0.6, 1.2, 1.6, 0.4]


Notice how the difference in the reseampled means is sometimes greater, and sometimes less? 

In [ ]:
def resample(l):
    final = []
    for i in range(len(l)):
        final.append(random.choice(l))